beefy-gadget/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+			return false;
+		}
+
+		self.votes.push(vote);
+		true
+	}
+
+	fn is_done(&self, threshold: usize) -> bool {
+		self.votes.len() >= threshold
+	}
+}
+
+fn threshold(voters: usize) -> usize {
+	let faulty = voters.saturating_sub(1) / 3;
+	voters - faulty
+}
+
+struct Rounds<Hash, Id, Signature> {
+	rounds: BTreeMap<Hash, RoundTracker<Id, Signature>>,
+	voters: Vec<Id>,
+}
+
+impl<Hash, Id, Signature> Rounds<Hash, Id, Signature>
+where
+	Hash: Ord,
+{
+	fn new(voters: Vec<Id>) -> Self {
+		Rounds {
+			rounds: BTreeMap::new(),
+			voters,
+		}
+	}
+}
+
+impl<Hash, Id, Signature> Rounds<Hash, Id, Signature>
+where
+	Hash: Ord,
+	Id: PartialEq,
+	Signature: PartialEq,
+{
+	fn add_vote(&mut self, round: Hash, vote: (Id, Signature)) -> bool {
+		self.rounds.entry(round).or_default().add_vote(vote)
+	}
+
+	fn is_done(&self, round: &Hash) -> bool {
+		self.rounds
+			.get(round)
+			.map(|tracker| tracker.is_done(threshold(self.voters.len())))
+			.unwrap_or(false)
+	}
+
+	fn drop(&mut self, round: &Hash) {
+		self.rounds.remove(round);
+	}
+}
+
+fn topic<Block: BlockT>() -> Block::Hash {
+	<<Block::Header as HeaderT>::Hashing as HashT>::hash("beefy".as_bytes())
+}
+
+#[derive(Debug, Decode, Encode)]
+struct VoteMessage<Hash, Id, Signature> {
+	block: Hash,
+	id: Id,
+	signature: Signature,
+}
+
+struct BeefyWorker<Block: BlockT, Id, Signature, FinalityNotifications> {
+	local_id: Id,
+	key_store: BareCryptoStorePtr,
+	rounds: Rounds<Block::Hash, Id, Signature>,
+	finality_notifications: FinalityNotifications,
+	gossip_engine: Arc<Mutex<GossipEngine<Block>>>,
+	best_finalized_block: NumberFor<Block>,
+	best_block_voted_on: NumberFor<Block>,
+}
+
+impl<Block, Id, Signature, FinalityNotifications>
+	BeefyWorker<Block, Id, Signature, FinalityNotifications>
+where
+	Block: BlockT,
+{
+	fn new(
+		local_id: Id,
+		key_store: BareCryptoStorePtr,
+		voters: Vec<Id>,
+		finality_notifications: FinalityNotifications,
+		gossip_engine: GossipEngine<Block>,
+		best_finalized_block: NumberFor<Block>,
+		best_block_voted_on: NumberFor<Block>,
+	) -> Self {
+		BeefyWorker {
+			local_id,
+			key_store,
+			rounds: Rounds::new(voters),
+			finality_notifications,
+			gossip_engine: Arc::new(Mutex::new(gossip_engine)),
+			best_finalized_block,
+			best_block_voted_on,
+		}
+	}
+}
+
+impl<Block, Id, Signature, FinalityNotifications>
+	BeefyWorker<Block, Id, Signature, FinalityNotifications>
+where
+	Block: BlockT,
+	Id: Codec + Debug + PartialEq + Public,
+	Signature: Codec + Debug + PartialEq + std::convert::TryFrom<Vec<u8>>,
+	FinalityNotifications: Stream<Item = FinalityNotification<Block>> + Unpin,
+{
+	fn should_vote_on(&self, number: NumberFor<Block>) -> bool {
+		use sp_runtime::traits::Saturating;
+		use sp_runtime::SaturatedConversion;
+
+		let diff = self.best_finalized_block.saturating_sub(self.best_block_voted_on);
+		let diff = diff.saturated_into::<u32>();
+		let next_power_of_two = 2u32.pow((diff as f64 / 2.0).log2().ceil() as u32);
```

Comment:
Why not `https://doc.rust-lang.org/std/primitive.u32.html#method.next_power_of_two`?
```
(diff / 2).next_power_of_two()
```